keep/providers/splunk_provider/splunk_provider.py

```diff
@@ -0,0 +1,154 @@
+
+
+@pydantic.dataclasses.dataclass
+class SplunkProviderAuthConfig:
+    username: str = dataclasses.field(
+        metadata={
+            "required": True,
+            "description": "Splunk Username",
+        }
+    )
+    password: str = dataclasses.field(
+        metadata={
+            "required": True,
+            "description": "Splunk Password",
+            "sensitive": True,
+        }
+    )
+
+    host: str = dataclasses.field(
+        metadata={
+            "description": "Splunk Host (default is localhost)",
+        },
+        default="localhost"
+    )
+    port: int = dataclasses.field(
+        metadata={
+            "description": "Splunkd Port (default is 8089)",
+        },
+        default=8089
+    )
+
+
+class SplunkProvider(BaseProvider):
+    """Pull alerts and query incidents from Splunk."""
+
+    PROVIDER_SCOPES = [
+        ProviderScope(
+            name="connect_to_client",
+            description="The user can connect to the client",
+            mandatory=True,
+            alias="Connect to the client",
+        )
+    ]
+
+    SEVERITIES_MAP = {
+        "1": AlertSeverity.LOW,
+        "2": AlertSeverity.INFO,
+        "3": AlertSeverity.WARNING,
+        "4": AlertSeverity.HIGH,
+        "5": AlertSeverity.CRITICAL,
+    }
+
+    def __init__(
+            self, context_manager: ContextManager, provider_id: str, config: ProviderConfig
+    ):
+        super().__init__(context_manager, provider_id, config)
+
+    def validate_scopes(self) -> dict[str, bool | str]:
+        try:
+            connect(username=self.authentication_config.username, password=self.authentication_config.password,
+                    host=self.authentication_config.host, port=self.authentication_config.port)
+            scopes = {
+                "connect_to_client": True,
+            }
+        except Exception as e:
+            self.logger.exception("Error validating scopes")
+            scopes = {
+                "connect_to_client": str(e),
+            }
+        return scopes
+
+    def validate_config(self):
+        self.authentication_config = SplunkProviderAuthConfig(
+            **self.config.authentication
+        )
+
+    def dispose(self):
+        """
+        No need to dispose of anything, so just do nothing.
+        """
+        pass
+
+    def setup_webhook(
+            self, tenant_id: str, keep_api_url: str, api_key: str, setup_alerts: bool = True
+    ):
+        self.logger.info("Setting up Splunk webhook on all Alerts")
+        creation_updation_kwargs = {
+            "actions": "webhook",
+            "action.webhook": "1",
+            "action.webhook.param.url": keep_api_url,
+        }
+        service = connect(username=self.authentication_config.username, password=self.authentication_config.password,
+                          host=self.authentication_config.host, port=self.authentication_config.port)
+        for saved_search in service.saved_searches:
```

Comment:
"Saved searches" is how you define alerts in Splunk Enterprise?

Reply:
Yup.